source/common/profiler/profiler.cc

```diff
@@ -41,6 +45,101 @@ bool Heap::stopProfiler() {
+    ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::main), warn,
+                        "Profile: tcmalloc heap profiler was started or profile file path is empty "
+                        "or output period is zero");
+    return false;
+  }
+
+  heap_file_path_ = std::string(path);
+  period_ = period;
+
+  ASSERT(period_.count() > 0);
+
+  // TODO(wbpcode): should we create new independent thread for this?
+  flush_timer_ = Event::MainDispatcherSingleton::get().createTimer([this]() {
+    auto profile = tcmalloc::MallocExtension::SnapshotCurrent(tcmalloc::ProfileType::kHeap);
+    absl::StatusOr<std::string> result = tcmalloc::Marshal(profile);
+
+    if (!result.ok()) {
+      ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::main), warn,
+                          "Profile: get nothing from tcmalloc profile");
+      return;
+    }
+    const std::string output_file_path = heap_file_path_ + std::to_string(next_file_id_++);
+    std::ofstream output_file(output_file_path, std::ios_base::binary);
+
+    if (!output_file.is_open()) {
+      ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::main), warn,
+                          "Profile: cannot open profile file: {}", output_file_path);
+      return;
+    }
+
+    ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::main), debug,
+                        "Profile: write heap profile result to: {}", output_file_path);
+    output_file << result.value();
+    output_file.close();
+
+    flush_timer_->enableTimer(period_);
+  });
+
+  flush_timer_->enableTimer(period_);
+  ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::main), info,
+                      "Profile: tcmalloc heap profiler is started");
+  heap_profile_started_ = true;
+  return true;
+}
+
+bool TcmallocHeapProfiler::stopProfiler() {
+  if (!heap_profile_started_) {
+    ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::main), warn,
+                        "Profile: tcmalloc heap profiler was stopped");
+
+    return false;
+  }
+
+  heap_file_path_ = "";
+  period_ = {};
+  flush_timer_->disableTimer();
+  flush_timer_.reset();
+  heap_profile_started_ = false;
+
+  ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::main), info,
+                      "Profile: tcmalloc heap profiler is stopped");
+
+  return true;
+}
+
+bool Cpu::profilerEnabled() { return false; }
+bool Cpu::startProfiler(const std::string&) { return false; }
+void Cpu::stopProfiler() {}
+
+bool Heap::profilerEnabled() { return true; }
+bool Heap::isProfilerStarted() { return tcmallocHeapProfiler().heapProfilerStarted(); }
+bool Heap::startProfiler(const std::string& path) {
+  // Output heap profile file every minute by default.
+  // TODO(wbpcode): make this period configurable.
+  return tcmallocHeapProfiler().startProfiler(path, std::chrono::milliseconds(60000));
```

Comment:
Why do we need every minute output? Does this match gperftools behavior?

Reply:
IMO it would be simpler to just do a new API for this and just make people hit it to get a dump? Then they can do whatever they want in a script. That would avoid a lot of complexity around all the timer stuff? WDYT?

/wait

Reply:
Yeah, this way would be very simple. I will try to land it. 😄

Reply:
> Why do we need every minute output? Does this match gperftools behavior?

There is some differences wich gperftools. We can only get heap profile by calling extension API munually. So a timer and perid is necessary.

But I will refactor it accoriding to comment of matt. 😄